File: go/binarySearch/33_Search_in_Rotated_Sorted_Array/33_Search_in_Rotated_Sorted_Array.go
package searchinrotatedsortedarray

func search(nums []int, target int) int {
	left, right := 0, len(nums)-1

	for left <= right {
		mid := (left + right) / 2
		if nums[mid] == target {
			return mid
		}

		// Are we in the left sorted array or right sorted array. How to decide?
		// If the middle element is > nums[left] then we are in left sorted array
		// in right sorted array all the elements will be less than nums[0]

		if nums[mid] >= nums[left] {
			if target > nums[mid] {
				left = mid + 1 // search in right portion
			} else if target < nums[mid] && target < nums[left] {
				left = mid + 1 // search in right portion
			} else {
				// if target is less than nums[mid] and target is greater than nums[left] search in left side
				right = mid - 1
			}
		} else { // nums[mid] < nums[left]. pivot happened here all the elements after pivot would be less than nums[left]
			if target < nums[mid] {
				right = mid - 1 // search in left portion
			} else if target > nums[mid] && target > nums[right] {
				right = mid - 1 // search in left portion
			} else {
				left = mid + 1 // search in right portion
			}
		}
	}

	return -1
}

File: c/33-Search-in-Rotated-Sorted-Array.c
int findPivotIndex(int* nums, int numsSize)
{
        int s = 0;
        int e = numsSize;
        
        while(s <= e){
            
            int m = s + (e - s)/2;
            
            if(m < e && nums[m] > nums[m+1]){
                return m;
            }
            else if(s < m && nums[m] < nums[m-1]){
                return m - 1;
            }
            else if(nums[s] < nums[m]){
                s = m + 1;
            }
            else if(nums[s] >= nums[m]){
                e = m - 1;
            }
        }
        
        // If the array is not rotated then last position will be the pivot element.
    
        return numsSize;
}

int binarySearch(int *nums, int s, int e, int target)
{
    while(s <= e){
            
            int m = s + (e - s)/2;
            
            if(nums[m] == target){
                return m;
            }
            else if(nums[m] < target){
                s = m + 1;
            }
            else{
                e = m - 1;
            }
        }
        
        return -1;
}

int search(int* nums, int numsSize, int target){
    
    int n = numsSize - 1;
        
    int pivotIndex = findPivotIndex(nums, n);

    int firstTry = binarySearch(nums, 0, pivotIndex, target);

    if(firstTry != -1){
        return firstTry;
    }

    return binarySearch(nums, pivotIndex + 1, n, target);
}